### ISSUE TITLE
`reportUnknownVariableType` false positive on dataclass field when type annotation is a wider type than what the `default_factory` returns

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#1502]
### Description

Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sAaKYfbVGAKFEimTFVW2QXxgAdhgB0cAEbICxUuQCycIkXxCA5ixYABeEjSYMLfVigAxMGABcLKLagS4IS7Hz8A2hgQg6qhAF0oAF56RmYAChhsYDgAV1QEAH1o-lIAT0CYVwQAShs7AGIoMABrTTt7OAAvZ0VlVTUPLx8hfyCQphgIqNj4pJQENIys3PLCgBVUomwi4CgAIgdKuYI8IkcBOB5UqBihYqEwAHchPNtCuwmpmfnF5fw8Ocz3AFU9g%2BO6V-2joT85oA)

```python
from dataclasses import dataclass, field
from collections.abc import Mapping

@dataclass
class Foo:
    bar: dict[str, int] = field(default_factory=dict)
    # ok

    baz: Mapping[str, int] = field(default_factory=dict)
    # Type of "baz" is partially unknown
    #    Type of "baz" is "dict[Unknown, Unknown]"
```
